hydromt/raster.py

```diff
@@ -2233,6 +2234,243 @@ def tile_window(shape, px):
+                np.arange(round(maxy, 2), round(miny, 2), -dx),
+            )
+            for l, u in lu:
+                col = int(round(odx + (l - minx) / dx, 2))
+                row = int(round(ody + (maxy - u) / dx, 2))
+                bounds = [
+                    l,
+                    u - dx,
+                    l + dx,
+                    u,
+                ]
+                yield Affine(px, 0, l, 0, -px, u), bounds, col, row
+
+        zoom_levels.sort()
+        if zoom_levels[-1] > min_lvl:
+            min_lvl = zoom_levels[-1]
+        zoom_levels = list(range(zoom_levels[0], min_lvl + 1, 1))
+        zoom_levels.reverse()
+
+        create_folder(root)
+
+        vrt_fn = None
+        nodata = self.nodata
+        zls = {}
+        for zl in zoom_levels:
+            dx = (20037508.34 * 2) / (2**zl)
+            px = dx / px_size
+
+            # read data from previous zoomlevel
+            if vrt_fn is not None:
+                obj = xr.open_dataarray(vrt_fn, engine="rasterio").squeeze(
+                    "band", drop=True
+                )
+
+            obj_bounds = obj.raster.bounds
+            obj_res = obj.raster.res[0]
+            obj_shape = obj.raster.shape
+
+            t_r = pyproj.Transformer.from_crs(
+                3857, obj.raster.crs.to_epsg(), always_xy=True
+            )
+
+            sd = join(root, f"{zl}")
+            create_folder(sd)
+            txt_path = join(sd, "filelist.txt")
+            file = open(txt_path, "w")
+
+            for transform, bounds, col, row in tile_window(zl, minx, miny, maxx, maxy):
+                ssd = join(sd, f"{col}")
+                create_folder(ssd)
+
+                ul = t_r.transform(bounds[0], bounds[3])
+                lr = t_r.transform(bounds[2], bounds[1])
+                out_bounds = [ul[0], lr[1], lr[0], ul[1]]
+
+                temp_bounds = [
+                    max(out_bounds[0], obj_bounds[0]),
+                    max(out_bounds[1], obj_bounds[1]),
+                    min(out_bounds[2], obj_bounds[2]),
+                    min(out_bounds[3], obj_bounds[3]),
+                ]
+
+                _window = [
+                    abs(x - y) / obj_res for x, y in zip(temp_bounds, obj_bounds)
+                ]
+
+                u = math.floor(round(_window[3], 2))
+                l = math.floor(round(_window[0], 2))
+                h = obj_shape[0] - u - math.floor(round(_window[1], 2))
+                w = obj_shape[1] - l - math.floor(round(_window[2], 2))
+
+                temp = obj[u : u + h, l : l + w]
+
+                tile = np.empty((px_size, px_size), np.float64)
+                tile[:] = nodata
+
+                temp_transform = Affine(
+                    obj_res,
+                    0,
+                    obj_bounds[0] + l * obj_res,
+                    0,
+                    -obj_res,
+                    obj_bounds[3] - u * obj_res,
+                )
+
+                rasterio.warp.reproject(
```

Comment:
Why do you use the rasterio reproject method instead of the raster.reproject method in the same script? That would avoid a lot of lines of code like creating a new tmp array below.

Reply:
See issue #442

Reply:
I'll will write this out in the issue when I reopen it. For now I think the implementation is fine as it is, but in the future it would indeed be nicer to use the raster.reproject method. However I'd like to solve issue #442 first instead more or less 'hacking' it.